src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs

```diff
@@ -165,33 +165,106 @@ public Vector3 Translation
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
             public static Impl operator *(in Impl left, in Impl right)
             {
-                Impl result;
+                if (Vector256.IsHardwareAccelerated)
+                {
+                    var lhs = Unsafe.BitCast<Impl, Vector512<float>>(left);
+
+                    var t0 = lhs.GetLower();
+                    var t1 = lhs.GetUpper();
+
+                    var a0 = Vector256.Shuffle(t0, Vector256.Create(0, 0, 0, 0, 4, 4, 4, 4));
+                    var a1 = Vector256.Shuffle(t1, Vector256.Create(0, 0, 0, 0, 4, 4, 4, 4));
+                    var b = Vector256.Create(right.X.AsVector128(), right.X.AsVector128());
+                    var c0 = a0 * b;
+                    var c1 = a1 * b;
+
+                    a0 = Vector256.Shuffle(t0, Vector256.Create(1, 1, 1, 1, 5, 5, 5, 5));
+                    a1 = Vector256.Shuffle(t1, Vector256.Create(1, 1, 1, 1, 5, 5, 5, 5));
+                    b = Vector256.Create(right.Y.AsVector128(), right.Y.AsVector128());
+                    var c2 = a0 * b;
+                    var c3 = a1 * b;
+
+                    a0 = Vector256.Shuffle(t0, Vector256.Create(2, 2, 2, 2, 6, 6, 6, 6));
+                    a1 = Vector256.Shuffle(t1, Vector256.Create(2, 2, 2, 2, 6, 6, 6, 6));
+                    b = Vector256.Create(right.Z.AsVector128(), right.Z.AsVector128());
+                    var c4 = a0 * b;
+                    var c5 = a1 * b;
+
+                    a0 = Vector256.Shuffle(t0, Vector256.Create(3, 3, 3, 3, 7, 7, 7, 7));
+                    a1 = Vector256.Shuffle(t1, Vector256.Create(3, 3, 3, 3, 7, 7, 7, 7));
+                    b = Vector256.Create(right.W.AsVector128(), right.W.AsVector128());
+
+                    var n0 = c0 + c2 + c4 + (a0 * b);
+                    var n1 = c1 + c3 + c5 + (a1 * b);
+
+                    var result = Vector512.Create(n0, n1);
+                    return Unsafe.BitCast<Vector512<float>, Impl>(result);
+                }
+                else if (Vector128.IsHardwareAccelerated)
```

Comment:
I don't think it's worth us having this other implementation here today.

It's somewhat "unnatural" since a `Matrix4x4` is logically `4x Vector4` and may hit other pessimizations due to aligned loads/stores, etc. All for a relatively minor perf increase which may not actually be "better" due to higher risk of hitting JIT inlining limitations.

-- Bitcasting to `Vector512` is also not really ideal here and may cause different perf/codegen on V512 accelerated platforms.

Reply:
The bitcast to `Vector512` is because I originally wrote it for `Vector512` before I split it up because basically no one has AVX512. What would be the preferred method of dissecting the matrix?

`Matrix4x4` multiplication is just like `Quaternion` multiplication an operation that tends to be in the hot path of 3d applications.

Reply:
Same as for quaternion, this isn't so hot nor is the new code that mmuch more performance.

The really hot code in 3D applications executes on the GPU. There is quite a substantial amount that may still be executed on the CPU, but the 1-3ns difference doesn't justify the significantly more complex implementation and in some cases may even make it slower due to missing guarantees around alignment or the like.

Reply:
I didn't mean 3d application as in games, but as in working in a 3d space. I found the bad Quaternion and Matrix4x4 multiplication speeds in dotnet8 while optimizing a robotics application with inherently long dependency chains unsuitable for GPU acceleration.